src/index.ts

```diff
@@ -22,37 +52,73 @@ export const goSync = <T, E extends Error>(fn: () => T): GoResult<T, E> => {
   }
 };
 
-const createGoError = <E extends Error>(err: unknown): GoResultError<E> => {
-  if (err instanceof Error) return fail(err);
-  return fail(new Error('' + err));
+const retryFnWrapper = async <T, E extends Error>(
+  fn: Promise<T> | (() => Promise<T>),
+  attemptOptions: AttemptOptions<T>
+): Promise<GoResult<T, E>> => {
+  if (typeof fn === 'function') {
+    return retryFn(fn, attemptOptions);
+  }
+  return retryFn(() => fn, attemptOptions);
+};
+
+const retryFn = async <T, E extends Error>(
+  fn: () => Promise<T>,
+  attemptOptions: AttemptOptions<T>
+): Promise<GoResult<T, E>> => {
+  return retry(fn, attemptOptions)
+    .then(success)
+    .catch((err) => {
```

Comment:
```
.catch(createGoError)
```

Reply:
hmm looks like we can do:
```js
.catch<GoResultError<E>>(createGoError);
```
or
```js
.catch((err) => createGoError(err));
```

Any preference? :smile:

Reply:
Oh, I haven't realized that TS will complain about that. I find the latter nicer, but this is not that important :)